Fix Aria scan for IN filters

mbasmanova opened the pull request:
This set of changes is fixing scan for NOT IN and IN with large number of values. TestHiveDistributedQueries#testLargeIn is now passing.

```
SELECT orderkey FROM orders WHERE orderkey NOT IN (1, 2, 3);

SELECT orderkey FROM orders WHERE orderkey IN (1, 2, 3,...5000);
```

oerling commented:
There could be a bug in a case where 

 

C is null or (c < 10 and c > 11)

 

The range is always false but the resulting filter would be always false instead of is null.

 

Is there a TupleDomain representation for c not in (1, 2, 3) ?

 

 

 

 

From: Maria Basmanova <notifications@github.com> 
Sent: Thursday, March 7, 2019 1:03 PM
To: prestodb/presto <presto@noreply.github.com>
Cc: oerling <erling@xs4all.nl>; Review requested <review_requested@noreply.github.com>
Subject: Re: [prestodb/presto] Fix Aria scan for NOT IN filter (#12445)

 

@mbasmanova <https://github.com/mbasmanova>  requested your review on: #12445 <https://github.com/prestodb/presto/pull/12445>  Fix Aria scan for NOT IN filter.

—
You are receiving this because your review was requested.
Reply to this email directly, view it on GitHub <https://github.com/prestodb/presto/pull/12445#event-2188452662> , or mute the thread <https://github.com/notifications/unsubscribe-auth/Ap73zw8mll5K6l5-e3Wzi_HMX1rpf56Hks5vUX5ogaJpZM4bkGip> .

mbasmanova commented:
@oerling 

> Is there a TupleDomain representation for c not in (1, 2, 3) ?

This is represented as c != 1 and c != 2 and c != 3, e.g. all - 1 - 2 -3. See com.facebook.presto.sql.planner.TestDomainTranslator#testInOptimization

```
        testDomain = Domain.create(
                ValueSet.ofRanges(
                        Range.lessThan(BIGINT, 4L))
                        .intersect(ValueSet.all(BIGINT)
                                .subtract(ValueSet.ofRanges(Range.equal(BIGINT, 1L), Range.equal(BIGINT, 2L), Range.equal(BIGINT, 3L))))
                        .union(ValueSet.ofRanges(Range.range(BIGINT, 7L, true, 9L, true))), false);

        tupleDomain = withColumnDomains(ImmutableMap.<Symbol, Domain>builder().put(C_BIGINT, testDomain).build());
        assertEquals(toPredicate(tupleDomain), or(and(lessThan(C_BIGINT, bigintLiteral(4L)), not(in(C_BIGINT, ImmutableList.of(1L, 2L, 3L)))), between(C_BIGINT, bigintLiteral(7L), bigintLiteral(9L))));
```

mbasmanova commented:
@oerling

> There could be a bug in a case where C is null or (c < 10 and c > 11)

I tested this and I saw that the filter was is-null.

mbasmanova commented:
@oerling 

> There could be a bug in a case where C is null or (c < 10 and c > 11)

This particular case works fine, but you are correct that the original version was losing `nullAllowed`. I fixed that:

```
                if (rangeFilters.isEmpty()) {
                        filter = nullAllowed ? Filters.isNull() : Filters.alwaysFalse();
                    }
                    else {
                        filter = Filters.createMultiRange(rangeFilters, nullAllowed);
                    }
```

elonazoulay commented on presto-main/src/main/java/com/facebook/presto/sql/gen/ExpressionCompiler.java:
```diff
@@ -277,43 +278,45 @@ else if (expression instanceof LambdaDefinitionExpression) {
+        }
+
+        for (List<RowExpression> conjuncts : inputsToConjuncts.values()) {
+            RowExpression firstConjunct = conjuncts.get(0);
```
Should there be a check that conjuncts is not empty?

mbasmanova replied:
@elonazoulay The above loop never creates empty lists. Are you suggesting to add a verify call here?

elonazoulay commented on presto-main/src/main/java/com/facebook/presto/sql/gen/ExpressionCompiler.java:
```diff
@@ -277,43 +278,45 @@ else if (expression instanceof LambdaDefinitionExpression) {
+        for (RowExpression conjunct : allConjuncts) {
+            ImmutableSet.Builder<InputReferenceExpression> inputs = ImmutableSet.builder();
+            collectInputs(conjunct, inputs);
+            inputsToConjuncts.compute(inputs.build(),
```
What do you think of this:
`inputsToConjuncts.computeIfAbsent(inputs.build(), k -> new ArrayList<>).add(conjunct);`

elonazoulay commented on presto-orc/src/main/java/com/facebook/presto/orc/Filters.java:
```diff
@@ -188,6 +224,39 @@ public boolean isEquality()
+                return true;
+            }
+
+            if (o == null || getClass() != o.getClass()) {
```
Why not use instanceof? It returns false if o == null:
`if (o instanceof BigIntRange)`

elonazoulay commented on presto-orc/src/main/java/com/facebook/presto/orc/Filters.java:
```diff
@@ -444,26 +513,29 @@ public void addMember(SubfieldPath.PathElement member, Filter filter)
-                        throw new IllegalArgumentException("Bigint filter range set must be in ascending order of lower bound and ranges must be disjoint");
-                    }
+            this.filters = filters.toArray(new Filter[0]);
+            if (filters.get(0) instanceof BigintRange) {
```
Should there be a not empty check for filters before the get(0)?

elonazoulay commented on presto-orc/src/main/java/com/facebook/presto/orc/Filters.java:
```diff
@@ -503,25 +575,59 @@ public boolean testBytes(byte[] buffer, int offset, int length)
+                return true;
+            }
+
+            if (o == null || getClass() != o.getClass()) {
```
Why not use instanceof here as well?

elonazoulay commented on presto-orc/src/main/java/com/facebook/presto/orc/Filters.java:
```diff
@@ -503,25 +575,59 @@ public boolean testBytes(byte[] buffer, int offset, int length)
-        static final long M = 0xc6a4a7935bd1e995L;
-        private long[] longs;
-        int size;
+        private static final long emptyMarker = 0xdeadbeefbadefeedL;
```
Is there a chance that a value can equal the emptyMarker?

mbasmanova replied:
@elonazoulay Yes, that's possible. There is extra logic to handle this case.

tdcmeehan replied:
nit: `EMPTY_MARKER`

mbasmanova replied:
@tdcmeehan Updated

elonazoulay commented on presto-orc/src/main/java/com/facebook/presto/orc/Filters.java:
```diff
@@ -557,12 +663,45 @@ public boolean testLong(long value)
+                return true;
+            }
+
+            if (o == null || getClass() != o.getClass()) {
```
Can you use instanceof here as well?

elonazoulay commented on presto-orc/src/main/java/com/facebook/presto/orc/TupleDomainOrcPredicate.java:
```diff
@@ -357,8 +357,16 @@ else if (ranges.size() == 1) {
+                    List<Filter> rangeFilters = ranges.stream()
+                            .map(r -> createRangeFilter(type, r, false))
+                            .filter(f -> f != Filters.alwaysFalse())
+                            .collect(toList());
```
.collect(toImmutableList())

tdcmeehan commented on presto-orc/src/main/java/com/facebook/presto/orc/Filters.java:
```diff
@@ -444,26 +513,29 @@ public void addMember(SubfieldPath.PathElement member, Filter filter)
+                        .toArray();
+
+                for (int i = 1; i < longLowerBounds.length; i++) {
+                    checkArgument(longLowerBounds[i] >= upperBounds[i - 1], "bigint ranges must not overlap");
```
They may not be in ascending order?

mbasmanova replied:
They shouldn't be.

tdcmeehan commented on presto-orc/src/test/java/com/facebook/presto/orc/TestTupleDomainOrcPredicateFilters.java:
```diff
@@ -0,0 +1,433 @@
+    }
+
+    @AfterClass(alwaysRun = true)
+    public void tearDown()
```
Why do we need to do this?

mbasmanova replied:
We don't. Removing.

oerling commented:
BigintRange could be declared final. Besides, if o were an instance of a subclass of BigintRange other than the class of this, the test would fail.

 

 

 

 

From: Timothy Meehan <notifications@github.com> 
Sent: Tuesday, March 12, 2019 9:16 AM
To: prestodb/presto <presto@noreply.github.com>
Cc: oerling <erling@xs4all.nl>; Mention <mention@noreply.github.com>
Subject: Re: [prestodb/presto] Fix Aria scan for NOT IN filter (#12445)

 

@tdcmeehan commented on this pull request.

  _____  

In presto-orc/src/main/java/com/facebook/presto/orc/Filters.java <https://github.com/prestodb/presto/pull/12445#discussion_r264758864> :

> @@ -188,6 +224,39 @@ public boolean isEquality()
         {
             return upper == lower;
         }
+
+        @Override
+        public boolean equals(Object o)
+        {
+            if (this == o) {
+                return true;
+            }
+
+            if (o == null || getClass() != o.getClass()) {

This is probably generated code, and I'm not sure if we want equals to return true if o is some theoretical subclass of BigIntRange

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub <https://github.com/prestodb/presto/pull/12445#discussion_r264758864> , or mute the thread <https://github.com/notifications/unsubscribe-auth/Ap73z82aSTaEsALr8jP5nXpiQUoilDE0ks5vV9LRgaJpZM4bkGip> .

mbasmanova commented:
@tdcmeehan @elonazoulay @oerling Folks, thank you for reviewing. I addressed the comments and added a commit that fixes queries with IN predicates with large number of values. Please, take another look.

mbasmanova commented on presto-spi/src/main/java/com/facebook/presto/spi/predicate/ValueSet.java:
```diff
@@ -65,7 +65,7 @@ static ValueSet of(Type type, Object first, Object... rest)
         throw new IllegalArgumentException("Cannot create discrete ValueSet with non-comparable type: " + type);
     }
 
-    static ValueSet copyOf(Type type, Collection<Object> values)
```
@tdcmeehan This didn't work, hence, I changed it. However, I see that I no longer use this method. I'll revert the change.
